File: Try_GameEngine5/src/com/example/try_gameengine/action/MovementActionItemForMilliseconds.java
package com.example.try_gameengine.action;

import java.util.ArrayList;
import java.util.List;

import com.example.try_gameengine.action.visitor.IMovementActionVisitor;
import android.os.CountDownTimer;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

/**
 * MovementActionItem is a item(leaf) in MovementAction composites.
 * 
 * @author irons
 * 
 */
public class MovementActionItemForMilliseconds extends MovementActionItem {
	long millisTotal;
	long millisDelay;
	float dx;
	float dy;
	long resumeTotal;
	long resetTotal;
	
	boolean isReset = true;
	boolean isActionFinish = false;
	public String name;
	public int frameIdx;
	public boolean isStop = false;
	boolean isFirstTime = true;
	/**
	 * constructor.
	 * 
	 * @param millisTotal
	 * 			milliseconds for whole action running.
	 * @param millisDelay
	 * 			milliseconds for delay.
	 * @param dx
	 * 			x-dir move for per delay time.
	 * @param dy
	 * 			y-dir move for per delay time.
	 */
	public MovementActionItemForMilliseconds(long millisTotal, long millisDelay, final int dx,
			final int dy) {
		this(millisTotal, millisDelay, dx, dy, "MovementItem");
	}

	/**
	 * constructor.
	 * 
	 * @param millisTotal
	 * 			milliseconds for whole action running.
	 * @param millisDelay
	 * 			milliseconds for delay.
	 * @param dx
	 * 			x-dir move for per delay time.
	 * @param dy
	 * 			y-dir move for per delay time.
	 * @param description
	 * 			description for this movement action.
	 */
	public MovementActionItemForMilliseconds(long millisTotal, long millisDelay, final int dx,
			final int dy, String description) {
		super(new MovementActionInfo(millisTotal, millisDelay, dx, dy, description));
	}

	/**
	 * constructor.
	 * 
	 * @param info
	 */
	public MovementActionItemForMilliseconds(MovementActionInfo info) {
		super(info);
	}

	@Override
	public void start() {
		// TODO Auto-generated method stub

	}

	@Override
	protected MovementAction initTimer(){ super.initTimer();
		return this;
	}

	@Override
	public MovementAction getAction() {
		return this;
	}

	@Override
	public List<MovementAction> getActions() {
		return actions;
	}

	@Override
	public MovementActionInfo getInfo() {
		return info;
	}

	@Override
	public void setInfo(MovementActionInfo info) {
		this.info = info;
	}

	@Override
	public List<MovementAction> getCurrentActionList() {
		List<MovementAction> actions = new ArrayList<MovementAction>();
		actions.add(this);
		return actions;
	}

	@Override
	public List<MovementActionInfo> getCurrentInfoList() {
		List<MovementActionInfo> infos = new ArrayList<MovementActionInfo>();
		infos.add(this.info);
		return infos;
	}

	@Override
	public List<MovementActionInfo> getMovementInfoList() {
		List<MovementActionInfo> infos = new ArrayList<MovementActionInfo>();
		infos.add(this.info);
		return infos;
	}

	@Override
	public void cancelMove() {
	}

	@Override
	void pause() {
	}

	@Override
	public boolean isFinish() {
		return isActionFinish;
	}
}

File: Try_GameEngine5/src/com/example/try_gameengine/action/MovementActionItemForMilliseconds.kt
package com.example.try_gameengine.action

/**
 * MovementActionItem is a item(leaf) in MovementAction composites.
 * 
 * @author irons
 // */
open class MovementActionItemForMilliseconds : MovementActionItem {
    var millisTotal: Long = 0
    var millisDelay: Long = 0
    @JvmField
    var dx: Float = 0f
    @JvmField
    var dy: Float = 0f
    var resumeTotal: Long = 0
    var resetTotal: Long = 0

    override var isReset: Boolean = true
    var isActionFinish: Boolean = false
    override var name: String? = null
    var frameIdx: Int = 0
    override var isStop: Boolean = false
    override var isFirstTime: Boolean = true
    /**
     * constructor.
     * 
     * @param millisTotal
     * milliseconds for whole action running.
     * @param millisDelay
     * milliseconds for delay.
     * @param dx
     * x-dir move for per delay time.
     * @param dy
     * y-dir move for per delay time.
     * @param description
     * description for this movement action.
     // */
    /**
     * constructor.
     * 
     * @param millisTotal
     * milliseconds for whole action running.
     * @param millisDelay
     * milliseconds for delay.
     * @param dx
     * x-dir move for per delay time.
     * @param dy
     * y-dir move for per delay time.
     // */
    @JvmOverloads
    constructor(
        millisTotal: Long, millisDelay: Long, dx: Int,
        dy: Int, description: String? = "MovementItem"
    ) : super(MovementActionInfo(millisTotal, millisDelay, dx.toFloat(), dy.toFloat(), description))

    /**
     * constructor.
     * 
     * @param info
     // */
    constructor(info: MovementActionInfo) : super(info)

    public override fun start() {
        // TODO Auto-generated method stub
    }

    public override fun initTimer(): MovementAction? {
        super.initTimer()
        return this
    }

    public override fun getAction(): MovementAction {
        return this
    }

    public override fun getActions(): MutableList<MovementAction> {
        return actions
    }

    public override fun getInfo(): MovementActionInfo {
        return info
    }

    public override fun setInfo(info: MovementActionInfo?) {
        this.info = info ?: return
    }

    public override fun getCurrentActionList(): MutableList<MovementAction> {
        val actions: MutableList<MovementAction> = ArrayList<MovementAction>()
        actions.add(this)
        return actions
    }

    public override fun getCurrentInfoList(): MutableList<MovementActionInfo?> {
        val infos: MutableList<MovementActionInfo?> = ArrayList<MovementActionInfo?>()
        infos.add(this.info)
        return infos
    }

    public override fun getMovementInfoList(): MutableList<MovementActionInfo?> {
        val infos: MutableList<MovementActionInfo?> = ArrayList<MovementActionInfo?>()
        infos.add(this.info)
        return infos
    }

    public override fun cancelMove() {
    }

    override fun pause() {
    }

    public override fun isFinish(): Boolean {
        return isActionFinish
    }

    @Throws(CloneNotSupportedException::class)
    public override fun clone(): MovementActionItemForMilliseconds {
        val copy = MovementActionItemForMilliseconds(this.info.clone())
        copy.actionListener = this.actionListener
        copy.timerOnTickListener = this.timerOnTickListener
        copy.controller = this.controller
        copy.timerOnTickListener = this.timerOnTickListener
        for (action in this.actions) {
            val subCopy = action.clone() as MovementAction
            copy.addMovementAction(subCopy)
        }
        copy.name = name
        return copy
    }
}
